### PR TITLE
kcptun: bump to v20190809, merge config and service scripts

### PR DESCRIPTION
**Pull request by expiron**
Maintainer: me / @liudf0716 
Compile tested: ar71xx, tl-wdr6500v5, OpenWrt SNAPSHOT r10750, Ubuntu 16.04 Host
Run tested: ar71xx, tl-wdr6500-v5, OpenWrt SNAPSHOT r10750-7ec092e641

This PR updates:
- bump version to v20190809
- fix `PKG_LICENSE_FILES`
- rename package names to `kcptun-server`, `kcptun-client`
- refactor Makefile
- merge config and service scripts
- add more config options to support all features
- add additional options: `gogc`, `syslog`, `user`
- add README.md

Reference:
- package shadowsocks-libev

### CODE REVIEW

**Comment by expiron** [comment]
It looks like something in sdk went wrong.
CircleCI log:
```
Updating feed 'base' from 'https://github.com/openwrt/openwrt.git;master' ...
Cloning into './feeds/base'...
remote: Enumerating objects: 9088, done.        
remote: Counting objects: 100% (9088/9088), done.        
remote: Compressing objects: 100% (7628/7628), done.        
Receiving objects: 100% (9088/9088), 12.33 MiB | 10.17 MiB/s, done.
remote: Total 9088 (delta 1625), reused 4799 (delta 828), pack-reused 0        
Resolving deltas: 100% (1625/1625), done.
Create index file './feeds/base.index' 
make: *** [staging_dir/host/.prereq-build] Error 1
Collecting target info: doneUpdating feed 'packages' from '/home/build/openwrt_packages' ...
Create index file './feeds/packages.index' 
make: *** [staging_dir/host/.prereq-build] Error 1
Collecting target info: doneUpdating feed 'luci' from 'https://github.com/openwrt/luci.git;master' ...
Cloning into './feeds/luci'...
Warning: Permanently added the RSA host key for IP address '140.82.113.4' to the list of known hosts.

remote: Enumerating objects: 4140, done.        
remote: Counting objects: 100% (4140/4140), done.        
remote: Compressing objects: 100% (2105/2105), done.        
remote: Total 4140 (delta 915), reused 2999 (delta 562), pack-reused 0
Receiving objects: 100% (4140/4140), 4.16 MiB | 0 bytes/s, done.
Resolving deltas: 100% (915/915), done.
Create index file './feeds/luci.index' 
make: *** [staging_dir/host/.prereq-build] Error 1
Collecting target info: donemake: *** [staging_dir/host/.prereq-build] Error 1
Exited with code 2
```

**Comment by neheb** [comment]
Please rebase against current master to fix the buildbot.

**Comment by expiron** [comment]
@neheb What should I do next?

File: net/kcptun/Makefile [hunk 1]
```diff
@@ -1,74 +1,65 @@
...
+$(eval $(call kcptun/templates))
 
-$(eval $(call GoBinPackage,kcptun-c))
-$(eval $(call BuildPackage,kcptun-c))
-$(eval $(call GoBinPackage,kcptun-s))
-$(eval $(call BuildPackage,kcptun-s))
+$(foreach component,$(KCPTUN_COMPONENTS), \
+  $(eval $(call GoBinPackage,kcptun-$(component))) \
+  $(eval $(call BuildPackage,kcptun-$(component))) \
+)
```
Comment (jefferyto) [inline net/kcptun/Makefile]: Why not include `$(call Package/kcptun/Default,$(component))` from `kcptun/templates` in this loop?

**Comment by jefferyto** [comment]
I think the major changes (rename packages, merge config and service scripts) will need to be reviewed by @liudf0716.

@expiron Can you explain why it would be better to have these changes?

File: net/kcptun/Makefile (hunk 1 above)
Comment (expiron) [reply net/kcptun/Makefile]: @jefferyto Fixed, thanks for your suggestion!

**Comment by expiron** [comment]
> @expiron Can you explain why it would be better to have these changes?

1. I think `kcptun-server` and `kcptun-client` are more explicit than `kcptun-s` and `kcptun-c`.
2. The same as package `shadowsocks-libev`, `kcptun-server` and `kcptun-client` have many common options. So their service scripts will have lots of similar parts. Merging into one script is necessary and good.

Additionally, I think "add more config options to support all features" is also a major change. But I can't make a brief enough title for them.

**Comment by liudf0716** [comment]
@expiron 
for me, I think kcptun-s &kcptun-c is more brief.
Merging two script and config into one is better, however it will bring more confusion to user, in addition, it will bring more bug and review to this PR

**Comment by liudf0716** [comment]
@expiron 
client and server using the same config file will cause the problem that client and server could not co-exist in the same device. do u think that reasonable?

**Comment by expiron** [comment]
@liudf0716 I don't think `kcptun-c` or `kcptun-s` can be easily recognized by a normal user who hasn't read your `Makefile`.
Imagine the scenery: someone is building a image using imagebuilder. If he wants to add kcptun client to his image, `kcptun-client` can be easily guessed but `kcptun-c` not.

**Comment by expiron** [comment]
> client and server using the same config file will cause the problem that client and server could not co-exist in the same device. do u think that reasonable?

@liudf0716 I don't think the problem you said exists. The service script will start all instances enabled in config file.
```
start_service() {
	local cfgtype

	mkdir -p "$confdir"
	config_load kcptun
	for cfgtype in server client; do
		config_foreach kcptun "$cfgtype" "$cfgtype"
	done
}
```
And the config file can contain `server` config and `client` config at the same time.

**Comment by liudf0716** [comment]
did u test install client and server pkg in your device? for me, install client first, then install server encounter error in my r7800

**Comment by expiron** [comment]
Due to my router only has 8MB flash, I only tested installing one package at the same time. It's my fault. Thanks for your further tests.
This problem has been fixed by adding a new package `kcptun-config` which contains config and service script. Both `kcptun-server` and `kcptun-client` depend on it.
Thanks a lot again! @liudf0716

**Comment by liudf0716** [comment]
@expiron **the following is just my opinion,** **maybe it's wrong**
firstly, kcptun client and server can co-exist in the same device, for example, for A and B , they all need to accelerate their own tcp service to each other, in this scene, kcptun client&server are totally different process, but in your package definition, kcptun client and server are bound, if I only want to change client's config, I need to restart both, for a product, such operation not wrong but appeared weird.
secondly,  setting config as independent package can resolve the co-exist problem, but for a product, it also appeared weird, In openwrt, there seems no such example, In industry, I don't think such design can be approved.

**Comment by liudf0716** [comment]
@expiron although I have above opinion, but I think your config script is more elegant in design and implementation

**Comment by expiron** [comment]
@liudf0716 
That's right to consider all aspects. I agree with your opinion about co-exist. But actually the service script can react as you expect in the scene you said before. I did the test in a openwrt docker container and got the [log](https://github.com/openwrt/packages/files/3514091/tests.log). You can find that only the instance associates changed config restarts. Maybe you just confused `service xxx reload` and `service xxx restart`.

Another I want to point out is that package [shadowsocks-libev](https://github.com/openwrt/packages/tree/master/net/shadowsocks-libev) is the "such example" in openwrt. In fact, what I have done in this PR is just learning from shadowsocks-libev and applying in kcptun.

**Comment by liudf0716** [comment]
@expiron 
If I only want to restart client without change config file, how to implement it?

**Comment by expiron** [comment]
Is it significant and necessary？If you have to do that, just kill the client.

**Comment by expiron** [comment]
@jefferyto  I think the discussion above has proved that this PR is reasonable and well-tested. Can it be merged?

**Comment by neheb** [comment]
A lot of shellcheck warnings show up. Most of them seem harmless. Up to @liudf0716 if he wants this in.

**Comment by liudf0716** [comment]
@neheb
I prefer to seperate kcptun client&server's config, anyway  this PR is ok

**Comment by jefferyto** [comment]
I don't really have a say in this, but I would prefer separate config files as well, rather than adding a package just to hold the config.

**Comment by expiron** [comment]
@BKPepe hello?

**Pull request merged**